Eclipsing object support

bmorris3 opened the pull request:
Here I add support for determining the observability of repeating events, with observations of transiting exoplanets and eclipsing binaries in mind. 

## API outline
Eclipsing systems are specified like this: 
```python
>>> from astropy.time import Time
>>> import astropy.units as u
>>> from astroplan import EclipsingSystem

>>> epoch = Time(2452826.628514, format='jd')
>>> period = 3.52474859 * u.day
>>> duration = 0.1277 * u.day

>>> hd209458 = EclipsingSystem(epoch=epoch, period=period, duration=duration,
                               name='HD 209458 b')
```
There are functions for computing the next eclipses of the system
```python
>>> observing_time = Time('2016-01-01 00:00')
>>> hd209458.next_primary_eclipse_time(observing_time)
<Time object: scale='utc' format='iso' value=['2016-01-03 16:16:09.848']>

>>> hd209458.next_secondary_eclipse_time(observing_time)
<Time object: scale='utc' format='iso' value=['2016-01-01 21:58:20.708']>
```
the next several eclipses of the system 
```python
>>> hd209458.next_primary_eclipse_time(observing_time, n_eclipses=5)
<Time object: scale='utc' format='iso' value=['2016-01-03 16:16:09.848' '2016-01-07 04:51:48.126'
                                              '2016-01-10 17:27:26.404' '2016-01-14 06:03:04.682'
                                              '2016-01-17 18:38:42.960']>
```
and the ingress/egress times for the system
```python
>>> hd209458.next_primary_ingress_egress_time(observing_time, n_eclipses=3)
<Time object: scale='utc' format='jd' value=[[ 2457391.11404175  2457391.24174175]
                                             [ 2457394.63879034  2457394.76649034]
                                             [ 2457398.16353893  2457398.29123893]]>
```

I also introduce a new function `is_event_observable`, which allows the user to compute whether or not a given eclipse event is observable given a list of constraints: 
```python
>>> from astroplan import PrimaryEclipseConstraint, is_event_observable, AltitudeConstraint
>>> constraints = [AltitudeConstraint(min=3*u.deg)]
>>> is_event_observable(constraints, apo, target, times=midtransit_times)
array([[ True, False,  True, ...,  True, False,  True, False]], dtype=bool)
```

### Docs

I've uploaded a static view of the docs, now including a tutorial for the new module, for you to review [here](http://staff.washington.edu/bmmorris/docs/PR315_docs/tutorials/eclipsing.html).

### Testing

The tests that I've added for these new functions test astroplan's results against results from the [transit prediction service from the Czech Astronomical Society Exoplanet Transit Database](http://var2.astro.cz/ETD/predictions.php), which provides an easy online interface for doing similar computations (specifically for exoplanets).

### Future work

I'm currently working on creating new exoplanet database querying methods in astroquery in astropy/astroquery#771. When that PR is merged, I'm likely to add some convenience methods to this astroplan module for quickly instantiating `EclipsingSystem` objects with planet periods/epochs grabbed via astroquery.

### Recon

I'm going to tag some people here who might use or have comments about this tool, to ask for comments. If you're busy or don't care to comment, feel free to ignore and un-watch this PR. @kweis @adrn @barentsen @ejeschke @rodluger @jlustigy @jlurie

barentsen commented:
Exoplanet transit observation planning would be a nice feature for astroplan.

Two items of feedback from me:
1. The phase and duration of a secondary eclipse is a function of eccentricity, periastron, and inclination, not just period and duration of the primary.  I would recommend to either make it very clear that the current implementation only applies to circular orbits, or preferably, support the equations for eccentric orbits.
2. The [Transit Service  of the NASA Exoplanet Archive](https://exoplanetarchive.ipac.caltech.edu/cgi-bin/TransitView/nph-visibletbls?dataset=transits) provides a transit planning service which can serve as inspiration.  For example, their "about" page documents how they deal with non-circular orbits.  They also have an API which can be useful for testing.

bsipocz commented:
I guess this PR now overrides https://github.com/astropy/astroplan/pull/120? 

Btw @barentsen's first point was already been raised there, too.

StuartLittlefair commented:
Another issue more pertinent to eclipsing binaries is that the timescale needs to be specified and respected in calculations. For example an ephemeris might be specified in BJD, and the calculations need to include light travel time corrections.

I’ll take a deeper look at the PR tomorrow. 

This might seem like nitpicking but eclipses of short period binaries can only last minutes, and the ~8 minute light travel time can make the difference between hitting the eclipse and missing it in your schedule...

bmorris3 commented:
Thanks for the tips @barentsen and @StuartLittlefair. I've added some big fat warnings to every eclipse time method which make it clear that barycentric corrections are not presently being computed – see updated docs [here](http://staff.washington.edu/bmmorris/docs/PR315_docs/tutorials/eclipsing.html).

I also implemented a slightly more sophisticated secondary eclipse calculator, which does the first-order approximation to the secondary eclipse time for small eccentricities, `secondary_eclipse_phase = 0.5 * (1 + 4/pi * e * cos(omega) )` where `e` is the orbital eccentricity and `omega` is the argument of periapsis, see [Wright (2010) Eqn 33](https://arxiv.org/abs/1001.2010) for reference.

StuartLittlefair commented on astroplan/eclipsing.py:
```diff
@@ -0,0 +1,295 @@
+        phase_array : `~numpy.ndarray`
+            Phase at each ``time``, on range [0, 1)
+        """
+        return ((time - self.epoch).to(u.day).value %
```
We then use some code like this here:

```
# some tests here to catch only one of observer and coord specified
if coord is not None and observer is not None self.kind is not None:
     time_corrected = time + time.light_travel_time(coord, kind=self.kind, location=observer.location)
else:
    time_corrected = time

scale_getter = attrgetter(self.timescale)
return ((scale_getter(time_corrected) - scale_getter(self.epoch)).to(u.day).value % 
             self.period.to(u.day).value) / self.period.to(u.day.value

StuartLittlefair commented on astroplan/eclipsing.py:
```diff
@@ -0,0 +1,295 @@
+            Phase at each ``time``, on range [0, 1)
+        """
+        return ((time - self.epoch).to(u.day).value %
+                self.period.to(u.day).value) / self.period.to(u.day).value
```
Another (simpler?) alternative is to use

```
phase = (scale_getter(time_corrected) - scale_getter(self.epoch)) / self.period
return phase - np.floor(phase).astype(int)
````

eteq commented on astroplan/periodic.py:
```diff
@@ -0,0 +1,295 @@
+        return np.logical_not(np.logical_or(self.in_primary_eclipse(time),
+                                            self.in_secondary_eclipse(time)))
+
+    def next_primary_eclipse_time(self, time, n_eclipses=1):
```
What about `next_eclipse_time_primary`?  Mostly for consistent tab completion, really...

bmorris3 replied:
I understand the tab completion motivation, but I think in this case most users should type `next_<tab>` and find their option. I'm very biased by my experience as a user here, but I think the flowing grammatical order makes more sense to me.

bsipocz replied:
I agree on this one with @bmorris3, one usually looks for either primary or secondaries rather than just any eclipse types (though with @eteq's suggestion, it would have been easier to have a convenience function of ``next_eclipse()`` that includes both primaries and secondaries.

eteq commented on astroplan/periodic.py:
```diff
@@ -0,0 +1,295 @@
+                         np.arange(n_eclipses) * self.period)
+        return eclipse_times
+
+    def next_secondary_eclipse_time(self, time, n_eclipses=1):
```
Similar as above but `next_eclipse_time_secondary`

eteq commented on astroplan/constraints.py:
```diff
@@ -906,6 +990,60 @@ def is_observable(constraints, observer, targets, times=None,
     return np.any(constraint_arr, axis=1)
 
 
+def is_event_observable(constraints, observer, target, times=None,
```
I really like how straightforward this is both as an API and in implementation, given the previous work (no action required, just "this is cool 😉").

eteq commented on astroplan/periodic.py:
```diff
@@ -0,0 +1,295 @@
+                self.period.to(u.day).value) / self.period.to(u.day).value
+
+
+class EclipsingSystem(PeriodicEvent):
```
How do you feel about renaming `EclipsingSystem` to something like `ApproximateEccentricEclipsingSystem`?  (One could imagine splitting that into `EclipsingSystem` as a superclass and `CircularEclipsingSystem` as a subclass, although I'm not sure that's necessary.) 

This might not make sense if your hope is to update the implementation later to be less of an approximation, but my read is that this approximation is intended to be a *permanent* "fast approximation" event.  That leaves room for future slower but more accurate calculations.

bmorris3 replied:
I am very biased, but I think the majority of eclipsing systems that are observed from the ground are eclipsing binaries and transiting exoplanets, which can be computed with sufficient accuracy for scheduling purposes in this "fast" implementation. I would be thrilled if someone came along and implement a more accurate version someday, but I think that this object will be the workhorse for 90+% of users, so I'm going to preserve the simple name.

eteq replied:
Hmm.  The hope is that this be not just for "observed from the ground".  But I see your point that it's not necessarily wise to make names that assume future work has already happened...

bsipocz commented on astroplan/periodic.py:
```diff
@@ -0,0 +1,295 @@
+
+        Returns
+        -------
+        primary_eclipses : `~astropy.time.Time` of shape (``n_eclipses``, 2)
```
typo, this should be "secondary_eclipses"

bsipocz commented on astroplan/periodic.py:
```diff
@@ -0,0 +1,295 @@
+        Returns
+        -------
+        primary_eclipses : `~astropy.time.Time` of shape (``n_eclipses``, 2)
+            Times of ingress and egress for the next ``n_eclipses`` primary
```
here, too
